libc/stdlib_alloc.c

```diff
@@ -0,0 +1,318 @@
+*/
+
+TEST_GROUP(stdlib_malloc);
+
+TEST_SETUP(stdlib_malloc)
+{
+}
+
+
+TEST_TEAR_DOWN(stdlib_malloc)
+{
+}
+
+TEST(stdlib_malloc, malloc_basic)
+{
+	void *ptr;
+	ptr = malloc(BLOCK_SIZE);
+	TEST_ASSERT_NOT_NULL(ptr);
+#ifndef phoenix  // Temporarily disabled on phoenix because issues: #660
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(BLOCK_SIZE, malloc_usable_size(ptr));
+#endif
+	memset(ptr, 0x5a, BLOCK_SIZE);
+	TEST_ASSERT_EACH_EQUAL_HEX8(0x5a, ptr, BLOCK_SIZE);
+	free(ptr);
+}
+
+TEST(stdlib_malloc, malloc_small)
+{
+	void *ptr;
+	ptr = malloc(1);
+	TEST_ASSERT_NOT_NULL(ptr);
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(1, malloc_usable_size(ptr));
+#endif
+	memset(ptr, 0x2d, 1);
+	TEST_ASSERT_EACH_EQUAL_HEX8(0x2d, ptr, 1);
+	free(ptr);
+}
+
+TEST(stdlib_malloc, malloc_large)
+{
+	void *ptr;
+	ptr = malloc(BLOCK_SIZE * LARGE_SIZE);
+	TEST_ASSERT_NOT_NULL(ptr);
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(BLOCK_SIZE * LARGE_SIZE, malloc_usable_size(ptr));
+#endif
+	memset(ptr, 0xb4, BLOCK_SIZE * LARGE_SIZE);
+	TEST_ASSERT_EACH_EQUAL_HEX8(0xb4, ptr, BLOCK_SIZE * LARGE_SIZE);
+	free(ptr);
+}
+
+TEST(stdlib_malloc, malloc_multiple)
+{
+	void *ptr0;
+	ptr0 = malloc(sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr0);
+	TEST_ASSERT_TRUE((uintptr_t)ptr0 % sizeof(int) == 0);
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(sizeof(int), malloc_usable_size(ptr0));
+#endif
+	memset(ptr0, 0x50, sizeof(int));
+	TEST_ASSERT_EACH_EQUAL_HEX8(0x50, ptr0, sizeof(int));
+
+	void *ptr1;
+	ptr1 = malloc(sizeof(float));
+	TEST_ASSERT_NOT_NULL(ptr1);
+	TEST_ASSERT_TRUE((uintptr_t)ptr1 % sizeof(float) == 0);
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(sizeof(float), malloc_usable_size(ptr1));
+#endif
+	memset(ptr1, 0xa0, sizeof(float));
+	TEST_ASSERT_EACH_EQUAL_HEX8(0xa0, ptr1, sizeof(float));
+
+	void *ptr2;
+	ptr2 = malloc(sizeof(char));
+	TEST_ASSERT_NOT_NULL(ptr2);
+	TEST_ASSERT_TRUE((uintptr_t)ptr2 % sizeof(char) == 0);
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(sizeof(char), malloc_usable_size(ptr2));
+#endif
+	memset(ptr2, 0x28, sizeof(char));
+	TEST_ASSERT_EACH_EQUAL_HEX8(0x28, ptr2, sizeof(char));
+
+	free(ptr0);
+	free(ptr1);
+	free(ptr2);
+}
+
+
+TEST(stdlib_malloc, malloc_zero)
+{
+	void *ptr;
+	ptr = malloc(0);
+#ifdef phoenix
+	TEST_ASSERT_NULL(ptr);
+#else
+	TEST_ASSERT_NOT_NULL(ptr);
+#endif
+	free(ptr);
+}
+
+TEST(stdlib_malloc, calloc_basic)
+{
+	int *ptr = calloc(BLOCK_SIZE, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EACH_EQUAL_INT(0, ptr, BLOCK_SIZE);
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(BLOCK_SIZE, malloc_usable_size(ptr));
+#endif
+	free(ptr);
+}
+
+TEST(stdlib_malloc, calloc_zero)
+{
+	int *ptr0 = calloc(0, sizeof(int));
+#ifdef phoenix
+	TEST_ASSERT_NULL(ptr0);
+#else
+	TEST_ASSERT_NOT_NULL(ptr0);
+#endif
+
+	int *ptr1 = calloc(BLOCK_SIZE, 0);
+#ifdef phoenix
+	TEST_ASSERT_NULL(ptr0);
+#else
+	TEST_ASSERT_NOT_NULL(ptr0);
+#endif
+
+	free(ptr0);
+	free(ptr1);
+}
+
+TEST(stdlib_malloc, calloc_one)
```

Comment:
We should also have a test that tries to alloc too much memory, e.g. `calloc(1, SIZE_MAX)`, `calloc(SIZE_MAX, SIZE_MAX)`, `calloc(SIZE_MAX, 1)` and check if errno is set to `ENOMEM`. It also applies to malloc and realloc.